### PR TITLE
Link properties/decorators view as a standalone form

### PR DESCRIPTION
**Pull request by Mati365**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/contributing/git-commit-message-convention.html))

Internal (link): Make link properties/decorators view a standalone form.

---

### Additional information

Related issue https://github.com/ckeditor/ckeditor5/issues/17230

It's part of improving linking experience initiative and implemented according to description in this [comment](https://github.com/ckeditor/ckeditor5/issues/17230#issuecomment-2476196426).

https://github.com/user-attachments/assets/004d1162-3900-4483-a785-95fce38b5f7b

### CODE REVIEW

File: packages/ckeditor5-link/lang/contexts.json [hunk 1]
```diff
@@ -4,6 +4,7 @@
 	"Link URL": "Label for the URL input in the Link URL editing balloon.",
 	"Link URL must not be empty.": "An error text displayed when user attempted to enter an empty URL.",
 	"Link image": "Label for the image link button.",
+	"Link properties": "Label for the link properties link popup title.",
```
Comment (niegowski) [inline packages/ckeditor5-link/lang/contexts.json:7]:
```suggestion
	"Link properties": "Label for the link properties link balloon title.",
```

File: packages/ckeditor5-link/src/linkediting.ts [hunk 2]
```diff
@@ -90,7 +90,7 @@ export default class LinkEditing extends Plugin {
 		editor.config.define( 'link', {
 			allowCreatingEmptyLinks: false,
 			addTargetToExternalLinks: false,
-			toolbar: [ 'linkPreview', '|', 'editLink', 'unlink' ]
+			toolbar: [ 'linkPreview', '|', 'editLink', 'linkProperties', 'unlink' ]
```
Comment (niegowski) [inline packages/ckeditor5-link/src/linkediting.ts:93]: There is no point in adding this button if the integrator does not specify manual decorators.

Comment (niegowski) [reply packages/ckeditor5-link/src/linkediting.ts:93]: I'd filter it out later near:

```ts
toolbarView.fillFromConfig( editor.config.get( 'link.toolbar' )!, editor.ui.componentFactory );
```

File: packages/ckeditor5-link/src/linkconfig.ts [hunk 3]
```diff
@@ -189,13 +189,14 @@ export interface LinkConfig {
...
 	 * * `'editLink'`,
+	 * * `'linkProperties'`
 	 * * `'unlink'`.
 	 *
 	 * The default configuration for link toolbar is:
 	 *
 	 * ```ts
 	 * const linkConfig = {
-	 * 	toolbar: [ 'linkPreview', '|', 'editLink', 'unlink' ]
+	 * 	toolbar: [ 'linkPreview', '|', 'editLink', 'linkProperties', 'unlink' ]
```
Comment (niegowski) [inline packages/ckeditor5-link/src/linkconfig.ts:199]: Describe it as depending on manual decorators.

File: packages/ckeditor5-link/src/linkui.ts [hunk 4]
```diff
@@ -734,6 +754,29 @@ export default class LinkUI extends Plugin {
 		this.formView!.enableCssTransitions();
 	}
 
+	/**
+	 * Adds the {@link #propertiesView} to the {@link #_balloon}.
+	 */
+	private _addPropertiesView(): void {
+		if ( !this.formView ) {
```
Comment (niegowski) [inline packages/ckeditor5-link/src/linkui.ts:761]:
```suggestion
		if ( !this.propertiesView ) {
```

File: packages/ckeditor5-link/src/linkui.ts [hunk 5]
```diff
@@ -769,19 +812,19 @@ export default class LinkUI extends Plugin {
 		this.selectedLinkableText = undefined;
 
 		if ( linkCommand.value !== undefined ) {
-			this._removeAdvancedView();
+			this._removePropertiesView();
```
Comment (niegowski) [inline packages/ckeditor5-link/src/linkui.ts:815]:
```suggestion
```

File: packages/ckeditor5-link/src/ui/linkpropertiesview.ts [hunk 6]
```diff
@@ -216,6 +216,7 @@ export default class LinkAdvancedView extends View {
 			const listItemView = new ListItemView( this.locale );
 
 			listItemView.children.add( item );
+			item.delegate( 'execute' ).to( this, 'property:toggle' );
```
Comment (niegowski) [inline packages/ckeditor5-link/src/ui/linkpropertiesview.ts:219]: This is not needed.

File: packages/ckeditor5-core/theme/icons/settings.svg [hunk 7]
```diff
@@ -1 +1 @@
-<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20"><path d="M3.734 6.375H1.75a.75.75 0 0 1 0-1.5h1.984a2.626 2.626 0 0 1 5.032 0h9.48a.75.75 0 0 1 0 1.5h-9.48a2.626 2.626 0 0 1-5.032 0Zm1.141-.75a1.375 1.375 0 1 1 2.75 0 1.375 1.375 0 0 1-2.75 0Zm11.388 8h1.983a.75.75 0 0 1 0 1.5h-1.983a2.626 2.626 0 0 1-5.033 0H1.75a.75.75 0 0 1 0-1.5h9.48a2.626 2.626 0 0 1 5.033 0Zm-1.142.75a1.375 1.375 0 1 1-2.75 0 1.375 1.375 0 0 1 2.75 0Z" clip-rule="evenodd"/></svg>
\ No newline at end of file
+<svg width="20" height="20" fill="none" xmlns="http://www.w3.org/2000/svg"><path fill-rule="evenodd" clip-rule="evenodd" d="M3.734 6.375H1.75a.75.75 0 0 1 0-1.5h1.984a2.626 2.626 0 0 1 5.032 0h9.48a.75.75 0 0 1 0 1.5h-9.48a2.626 2.626 0 0 1-5.032 0Zm1.141-.75a1.375 1.375 0 1 1 2.75 0 1.375 1.375 0 0 1-2.75 0ZM16.263 14.625h1.983a.75.75 0 0 1 0 1.5h-1.983a2.626 2.626 0 0 1-5.033 0H1.75a.75.75 0 0 1 0-1.5h9.48a2.626 2.626 0 0 1 5.033 0Zm-1.142.75a1.375 1.375 0 1 1-2.75 0 1.375 1.375 0 0 1 2.75 0Z" fill="#000"/></svg>
```
Comment (pszczesniak) [inline packages/ckeditor5-core/theme/icons/settings.svg:1]:
```suggestion
<svg viewBox="0 0 20 20" xmlns="http://www.w3.org/2000/svg"><path fill-rule="evenodd" clip-rule="evenodd" d="M3.734 6.375H1.75a.75.75 0 0 1 0-1.5h1.984a2.626 2.626 0 0 1 5.032 0h9.48a.75.75 0 0 1 0 1.5h-9.48a2.626 2.626 0 0 1-5.032 0Zm1.141-.75a1.375 1.375 0 1 1 2.75 0 1.375 1.375 0 0 1-2.75 0ZM16.263 14.625h1.983a.75.75 0 0 1 0 1.5h-1.983a2.626 2.626 0 0 1-5.033 0H1.75a.75.75 0 0 1 0-1.5h9.48a2.626 2.626 0 0 1 5.033 0Zm-1.142.75a1.375 1.375 0 1 1-2.75 0 1.375 1.375 0 0 1 2.75 0Z"/></svg>
```

Please check how it works with :point\_up: 

Ive removed the fixed `height`, `width` and `fill` attributes and set the `viewBox`.

File: packages/ckeditor5-link/src/linkui.ts [hunk 8]
```diff
@@ -392,21 +388,20 @@ export default class LinkUI extends Plugin {
...
 		this.listenTo( view, 'back', () => {
-			// Make sure the focus always gets back to the editable _before_ removing the focused form view.
-			// Doing otherwise causes issues in some browsers. See https://github.com/ckeditor/ckeditor5-link/issues/193.
-			editor.editing.view.focus();
+			this._removePropertiesView();
 
-			this._removeAdvancedView();
-			this.formView!.focus();
+			// Move focus back to the editing view to prevent from losing it while current view is removed.
+			editor.editing.view.focus();
```
Comment (niegowski) [inline packages/ckeditor5-link/src/linkui.ts:404]: Maybe this should be called before we remove the view.

File: packages/ckeditor5-link/src/linkui.ts [hunk 9]
```diff
@@ -556,6 +552,30 @@ export default class LinkUI extends Plugin {
...
+				label: t( 'Link properties' ),
+				icon: icons.settings,
+				tooltip: true
+			} );
+
+			button.bind( 'isEnabled' ).to(
+				linkCommand, 'isEnabled',
+				linkCommand, 'manualDecorators',
+				( isEnabled, manualDecorators ) => isEnabled && manualDecorators.length > 0
+			);
```
Comment (niegowski) [inline packages/ckeditor5-link/src/linkui.ts:571]: As this component could be added to the main toolbar (not only to the link toolbar) it should also be disabled when the link command value is empty - the selection is not on the link.

File: packages/ckeditor5-link/src/ui/linkpropertiesview.ts [hunk 10]
```diff
@@ -33,9 +33,9 @@ import '../../theme/linkform.css';
 /**
  * The link form view controller class.
```
Comment (niegowski) [inline packages/ckeditor5-link/src/ui/linkpropertiesview.ts:34]: Please update this line.

File: packages/ckeditor5-link/tests/linkui.js [hunk 11]
```diff
@@ -32,14 +32,15 @@ import LinkUI from '../src/linkui.js';
 import LinkFormView from '../src/ui/linkformview.js';
 import LinkButtonView from '../src/ui/linkbuttonview.js';
 import LinkBookmarksView from '../src/ui/linkbookmarksview.js';
-import LinkAdvancedView from '../src/ui/linkadvancedview.js';
 import LinkPreviewButtonView from '../src/ui/linkpreviewbuttonview.js';
+import LinkPropertiesView from '../src/ui/linkpropertiesview.js';
 import { MenuBarMenuListItemButtonView, ToolbarView } from '@ckeditor/ckeditor5-ui';
 
 import linkIcon from '../theme/icons/link.svg';
+import ManualDecorator from '../src/utils/manualdecorator.js';
```
Comment (niegowski) [inline packages/ckeditor5-link/tests/linkui.js:40]: Please move this import before the svg import.

File: scripts/clean-up-svg-icons.mjs [hunk 12]
```diff
@@ -46,6 +46,7 @@ import { execSync } from 'child_process';
 // A list of icons that should not NOT be cleaned up. Their internal structure should not be changed
 // because, for instance, CSS animations may depend on it.
 const EXCLUDED_ICONS = [
+	'settings.svg',
```
Comment (niegowski) [inline scripts/clean-up-svg-icons.mjs:49]: Is this required? Can we generate the icon so it is safe to be "cleaned up"?

Comment (Mati365) [reply scripts/clean-up-svg-icons.mjs:49]: Nope, it don't work. Compressor breaks SVG icon.

image

File: packages/ckeditor5-link/src/linkui.ts [hunk 13]
```diff
@@ -211,7 +211,7 @@ export default class LinkUI extends Plugin {
 	private _createViews() {
 		this.toolbarView = this._createToolbarView();
 		this.formView = this._createFormView();
-		this.advancedView = this._createAdvancedView();
+		this.propertiesView = this._createPropertiesView();
```
Comment (pszczesniak) [inline packages/ckeditor5-link/src/linkui.ts:214]: This view is created always, even when there are no decorators defined.

Comment (Mati365) [reply packages/ckeditor5-link/src/linkui.ts]: Fixed https://github.com/ckeditor/ckeditor5/pull/17534/commits/d6e071dabce4307821cf8f9081b33558c48d7f25

**Review by pszczesniak — APPROVED** [review]
For me it looks ok :+1:

**Pull request merged**